### PR TITLE
Add support for dynamic mana infusion recipes

### PR DESCRIPTION
**Pull request by eutro**
This PR adds support for adding dynamic mana infusion recipes, i.e. recipes whose outputs depend on their inputs, for both CraftTweaker and API users.

Example:
Enchanted mana helment infusion recipe which transfers NBT.
before
after
With the script:
```zs
import crafttweaker.api.item.IItemStack;

val copyNbt = (output as IItemStack, input as IItemStack) => { return output.withTag(input.tag); };

<recipetype:botania:mana_infusion>.addRecipe("mana_infusion_helmet", <item:botania:manasteel_helmet>, <item:minecraft:iron_helmet>.anyDamage(), 500, null, copyNbt);
<recipetype:botania:mana_infusion>.addRecipe("mana_infusion_chestplate", <item:botania:manasteel_chestplate>, <item:minecraft:iron_chestplate>.anyDamage(), 800, null, copyNbt);
<recipetype:botania:mana_infusion>.addRecipe("mana_infusion_leggings", <item:botania:manasteel_leggings>, <item:minecraft:iron_leggings>.anyDamage(), 700, null, copyNbt);
<recipetype:botania:mana_infusion>.addRecipe("mana_infusion_boots", <item:botania:manasteel_boots>, <item:minecraft:iron_boots>.anyDamage(), 400, null, copyNbt);
```

### CODE REVIEW

File: src/main/java/vazkii/botania/api/recipe/IManaInfusionRecipe.java [hunk 1]
```diff
@@ -27,6 +27,15 @@
 
 	boolean matches(ItemStack stack);
 
+	@Nonnull
+	default ItemStack getRecipeOutputDynamic(@Nonnull ItemStack input) {
```
Comment (Hubry) [inline src/main/java/vazkii/botania/api/recipe/IManaInfusionRecipe.java:31]: Just call it `getRecipeOutput`, just like the elven trade recipe has one method with no params and one with a list of stacks.

**Review by Hubry — APPROVED** [review]

**Comment by Hubry** [comment]
Please fix conflicts, I touched the recipe manager :V

File: src/main/java/vazkii/botania/api/recipe/IManaInfusionRecipe.java [hunk 2]
```diff
@@ -25,15 +25,51 @@
...
+	@Nonnull
+	default ItemStack getRecipeOutput(@Nonnull ItemStack input) {
+		return getRecipeOutput().copy();
+	}
+
+	/**
+	 * Get the block state that must be under the Mana Pool for this recipe, or {@code null} if it can be anything.
+	 *
+	 * @return The catalyst block state.
+	 */
```
Comment (Hubry) [inline src/main/java/vazkii/botania/api/recipe/IManaInfusionRecipe.java:62]: This javadoc needs an update as this is using state ingredients now

**Review by Hubry — COMMENTED** [review]
Minor nitpick time :wacko:

File: src/main/java/vazkii/botania/api/recipe/IManaInfusionRecipe.java (hunk 2 above)
Comment (eutro) [reply src/main/java/vazkii/botania/api/recipe/IManaInfusionRecipe.java:62]: wack angery

**Review by Hubry — APPROVED** [review]

**Pull request merged**